### PR TITLE
Add severity to linter rules. Add new linter rules.

### PR DESCRIPTION
**Pull request by adewaleo**
- [x] Add `--min-severity` which specifies the minimum severity of the linter. Rules with lower severity will not be executed. This way we can distinguish between high priority rule infractions and lower priority infractions that could be helpful in command reviews. See `--severity` help text for `azdev linter`
- [x] Add `parameter_should_not_end_with_resource_group` rule - the only param that should end in `resource-group` is `--resource-group` --- **severity high**
- [x] Add `no_positional_parameters` rule - we frown on positional params in the CLI. --- **severity high**
- [x] Add `no_parameter_defaults_for_update_commands` rule - update command params should not have default values. This is so that they behave as no-ops when not set. --- **severity high**
- [x] Add `update_commands_support_generic_update`  rule - all update commands should expose generic update arguments. --- **severity high**
- [x] Command command loaders must have an associated resource type --- **severity medium**
- [x] expose wait command if a command in a group has --no-wait --- **severity medium**
- [x] location should not be required in a command. One can always default to resource group's location. --- **severity medium**
- [x] Delete commands of top level resources should prompt user for confirmation or accept --yes (not necessary if the resource is not a top level resource.). --- **severity medium**
- [x] arguments that end in `-id` smust be guids / uuids and not resource ids. This involves guessing from help text, and so this should be a medium or even low severity as we might not even be sure when warning the user. --- **severity low**

Todo: IF necessary, add other rules to assist in command reviews and to maintain CLI command / param / help quality.

such as:
- create / list commands should not have `--ids`. We have a check for list commands and its not possible to expose --ids on create commands. We might need to scrap this. --- **severity medium**. 
- When specifying parameter choices, they should use sdk enums [and get_enum_type). This might not be possible as is. it might be possible to detect this by adding a flag when registering arguments during command load, but that might not be worth it. --- **severity medium**

- Generally, an argument that accepts only one value should not be required, but default to the single value. An exception is in the event where it is possible / expected that this argument will take multiple values in the future, then it is fair not to default to a value. --- **severity low**

### CODE REVIEW

**Comment by adewaleo** [comment]
closed #119 for this so that changes can be made by the team to this branch.

The linter should be good to go. Severity works and the rules have been added. None of the old rules should be affected.

However, you might want to make changes or add exclusions before merging to master. #121 is a pr from an upstream branch, so anyone should be able to make changes to that PR / branch.

**Comment by adewaleo** [comment]
@MyronFanQiu @zikalino before this is merged to dev, the violations must be fixed or excluded with [linter_exclusions.yml](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/network/linter_exclusions.yml) files. See exclusion format here [linter_exclusions.yml](https://github.com/Azure/azure-cli/blob/7e28c54508e2ce3594db2988d864d84919005657/tools/automation/cli_linter/rules/linter_exclusions.yml).

please run azdev linter --ci to run the linter (defaults to high severity issues alone).

We don't need to address medium and low issues foe now as they are intended to help with command reviews. They shouldn't be run in CI.

**Comment by adewaleo** [comment]
Seems like [azdev's linter test ](https://dev.azure.com/azure-sdk/public/_build/results?buildId=81795&view=logs&j=aab6da13-c31e-5968-a14c-7ad7ef58e5d9&t=30d83fad-940a-5301-5843-ad66a693fb9c&l=193)the linter by running `azdev linter redis` on the CLI's source . Do we want to keep it this way or run the linter on all commands. This might make azdev's CI tests more tightly coupled to the CLI's CI

**Comment by MyronFanQiu** [comment]
> Seems like [azdev's linter test ](https://dev.azure.com/azure-sdk/public/_build/results?buildId=81795&view=logs&j=aab6da13-c31e-5968-a14c-7ad7ef58e5d9&t=30d83fad-940a-5301-5843-ad66a693fb9c&l=193)the linter by running `azdev linter redis` on the CLI's source . Do we want to keep it this way or run the linter on all commands. This might make azdev's CI tests more tightly coupled to the CLI's CI

@adewaleo  I prefer running the linter on all commands. Thanks. I think you can leave this one to me to fix so that I can ramp myself up.

File: azdev/operations/linter/linter.py [hunk 1]
```diff
@@ -228,17 +259,35 @@ def run(self, run_params=None, run_commands=None, run_command_groups=None, run_h
...
+                    violations = sorted(rule_func()) or []
+                    if violations:
+                        if rule_severity == LinterSeverity.HIGH:
+                            sev_color = Fore.RED
+                        elif rule_severity == LinterSeverity.MEDIUM:
+                            sev_color = Fore.YELLOW
+                        else:
+                            sev_color = Fore.CYAN
+
+                        print('- {} FAIL{} - {}{}{} severity: {}'.format(Fore.RED, Fore.RESET, sev_color,
```
Comment (MyronFanQiu) [inline azdev/operations/linter/linter.py]: Hello. Just want to know, will `print` directly show the log onto the screen? Do `Fore.RED` and `Fore.RESET` work like a switch?

**Comment by MyronFanQiu** [comment]
@adewaleo Hello, do you know how to run the azdev from source code? I don't find any doc introducing it.

**Comment by adewaleo** [comment]
> @adewaleo Hello, do you know how to run the azdev from source code? I don't find any doc introducing it. Based on the DevOps Pipelines, it seems I could download the artifacts and install it into a virtual env.

@MyronFanQiu you should be able to do so by running 

`pip install -e .` / `pip install -e /path/to/azdev/repo` at the root of your cloned local copy of the azdev source code, while your virtualenv is activated.

Make sure to first run pip install azdev to uninstall the public version of azdev.

**Comment by adewaleo** [comment]
> > Seems like [azdev's linter test ](https://dev.azure.com/azure-sdk/public/_build/results?buildId=81795&view=logs&j=aab6da13-c31e-5968-a14c-7ad7ef58e5d9&t=30d83fad-940a-5301-5843-ad66a693fb9c&l=193)the linter by running `azdev linter redis` on the CLI's source . Do we want to keep it this way or run the linter on all commands. This might make azdev's CI tests more tightly coupled to the CLI's CI
> 
> @adewaleo I prefer running the linter on all commands. Thanks. I think you can leave this one to me to fix so that I can ramp myself up. Then I will request your review later.

So just to clarify, we do run the linter on all commands in the CLI CI. Its up to you if you want to run it on all commands here. It might make it harder to commit linter changes here before updating CLI source code to conform to those changes.

**Comment by MyronFanQiu** [comment]
> > > Seems like [azdev's linter test ](https://dev.azure.com/azure-sdk/public/_build/results?buildId=81795&view=logs&j=aab6da13-c31e-5968-a14c-7ad7ef58e5d9&t=30d83fad-940a-5301-5843-ad66a693fb9c&l=193)the linter by running `azdev linter redis` on the CLI's source . Do we want to keep it this way or run the linter on all commands. This might make azdev's CI tests more tightly coupled to the CLI's CI
> > 
> > 
> > @adewaleo I prefer running the linter on all commands. Thanks. I think you can leave this one to me to fix so that I can ramp myself up. Then I will request your review later.
> 
> So just to clarify, we do run the linter on all commands in the CLI CI. Its up to you if you want to run it on all commands here. It might make it harder to commit linter changes here before updating CLI source code to conform to those changes. I leave it to you. I suppose there are pros and cons of doing it either way.

I see. Thanks for the explanation.

**Comment by yonzhan** [comment]
add to S169

**Comment by yonzhan** [comment]
S170

File: HISTORY.rst [hunk 2]
```diff
@@ -2,6 +2,13 @@
 
 Release History
 ===============
+0.1.14
```
Comment (haroldrandom) [inline HISTORY.rst:5]: should not update this, how about moving this to another PR while bumping a newer version?

File: setup.py [hunk 3]
```diff
@@ -77,6 +77,7 @@
     ],
     extras_require={
         ":python_version<'3.0'": ['pylint==1.9.2', 'futures'],
+        ":python_version<'3.4'": ['enum34'],
```
Comment (haroldrandom) [inline setup.py:80]: why adding this?

File: azdev/operations/linter/linter.py [hunk 4]
```diff
@@ -175,7 +208,8 @@ def get_linter():
                     return Linter(command_loader=command_loader, help_file_entries=help_file_entries,
                                   loaded_help=self._loaded_help)
                 return self.linter
-            self._rules[rule_type][rule_name] = rule_callable, get_linter
+
+            self._rules[rule_type][rule_name] = rule_callable, get_linter, rule_severity
```
Comment (haroldrandom) [inline azdev/operations/linter/linter.py:212]: Do you mean `(rule_callable, get_linter, rule_severity)`?

File: azdev/operations/linter/linter.py [hunk 5]
```diff
@@ -223,22 +257,65 @@ def run(self, run_params=None, run_commands=None, run_command_groups=None, run_h
...
         if not self.exit_code:
             print(os.linesep + 'No violations found.')
+
+        if self._update_global_exclusion is not None:
+            repo_paths = None
+            if self._update_global_exclusion == 'main':
+                repo_paths = [get_cli_repo_path()]
+            else:
+                repo_paths = get_ext_repo_paths()
+            exclusion_paths = [os.path.join(repo_path, 'linter_exclusions.yml') for repo_path in repo_paths]
```
Comment (haroldrandom) [inline azdev/operations/linter/linter.py:267]: edge case: If user don't `azdev setup` with `-r`, the `repo_paths` would be `None`, then it would be not iterable, such that this azdev would crash in this line.

File: azdev/operations/linter/linter.py [hunk 6]
```diff
@@ -223,22 +257,65 @@ def run(self, run_params=None, run_commands=None, run_command_groups=None, run_h
 
         if not self.exit_code:
             print(os.linesep + 'No violations found.')
+
+        if self._update_global_exclusion is not None:
+            repo_paths = None
```
Comment (haroldrandom) [inline azdev/operations/linter/linter.py:262]: No need to declare this variable like C++

File: azdev/operations/linter/linter.py (hunk 1 above)
Comment (haroldrandom) [reply azdev/operations/linter/linter.py:295]: Is there a conclusion?

File: azdev/operations/linter/rule_decorators.py [hunk 7]
```diff
@@ -4,54 +4,81 @@
...
 # -----------------------------------------------------------------------------
 
-from .linter import RuleError
+from knack.util import CLIError
+from .linter import RuleError, LinterSeverity
 
 
-def help_file_entry_rule(func):
-    return _get_decorator(func, 'help_file_entries', 'Help-Entry: `{}`')
+class AbstractRule(object):
```
Comment (haroldrandom) [inline azdev/operations/linter/rule_decorators.py:11]: I prefer not name it `Abstract` because it did't give any constraint to the subclass, it doesn't have any `@abstractmethod` decorator to the instance method.
How about 
```suggestion
class BaseRule:
```

File: azdev/operations/linter/linter.py (hunk 1 above)
Comment (MyronFanQiu) [reply azdev/operations/linter/linter.py:295]: yes.

File: azdev/operations/linter/linter.py (hunk 4 above)
Comment (MyronFanQiu) [reply azdev/operations/linter/linter.py:212]: yes! But looks like we don't need ()

File: setup.py (hunk 3 above)
Comment (MyronFanQiu) [reply setup.py:80]: removed
